src/python/pygrackle/utilities/grdata.py

```diff
@@ -0,0 +1,768 @@
+ └── user-data/                          # <- reserved for user data
+      ├── ...
+      └── ...
+[[[END:DIRECTORY-CARTOON]]]
+
+[[[BEGIN-SECTION:INTERNALS FILE REGISTRY]]]
+The file registry refers to a small file that maps the names of Grackle's
+datafiles with a checksum. At the time of writing we use sha1 checksums.
+
+In the long term, the registry is important in 2 -ish contexts:
+
+1. Embed the contents of the registry within the core Grackle library in order
+   to provide the option to directly read in files from the data-store without
+   being provided a full path. This embedding is handled by the build-system.
+   PR #237 proposes an implmentation for this logic.
+
+2. It's important for making make this tool work when:
+
+   - this tool is part of the pygrackle module.
+   - this tool is run as a portable single-file executable.
+[[[END-SECTION:INTERNALS-REGISTRY]]]
+"""
+
+# For use as a standalone command line tool this MUST ONLY use python's built-in modules
+import argparse
+import contextlib
+import filecmp
+import functools
+import hashlib
+import io
+import logging
+from math import log10
+import os
+import re
+import shutil
+import stat
+import sys
+import traceback
+from typing import IO, NamedTuple, Union
+import urllib.request
+from urllib.error import URLError, HTTPError
+
+
+if sys.version_info < (3, 6, 1):  # 3.6.0 doesn't support all NamedTuple features
+    raise RuntimeError("python 3.6.1 or newer is required")
+
+_CHUNKSIZE = 8192  # default chunksize used for file operations
+
+logger = logging.getLogger("grdata")
+logger.setLevel(logging.DEBUG)
+logger.addHandler(logging.StreamHandler())
+
+
+class GrdataError(RuntimeError):
+    pass
+
+
+@contextlib.contextmanager
+def _file_openner(f, mode, **kwargs):
+    if isinstance(f, (str, bytes, os.PathLike)):
+        with open(f, mode, **kwargs) as file:
+            yield file
+    else:
+        yield f
+
+
+def _progress_bar(total_bytes, silent=False):
+    """provides a function for drawing/updating progress bars"""
+    ncols = shutil.get_terminal_size()[0] - 1
+    power_div_3 = int(log10(total_bytes) // 3) if total_bytes > 0 else 0
+    factor, unit = 1000.0**power_div_3, (" B", "KB", "MB", "GB")[power_div_3]
+    # the output line has the form: '[<progress-bar>] <size>/<size> <unit>'
+    fmt = "\r[{bar:{barlen}.{nfill}}] {size:.2f}" + f"/{total_bytes/factor:.2f} {unit}"
+    barlen = ncols - 19  # for context, 15 <= (len(fmt.format(...)) - barlen) <= 19
+    suppress = (barlen < 1) or silent or not sys.stdout.isatty()
+    bar = None if suppress else (barlen * "=")
+
+    def _update(size):
+        nonlocal bar
+        if size is None and bar is not None:
+            print(flush=True)
+            bar = None
+        elif bar is not None:
+            nfill = int(barlen * (size / total_bytes))
+            val = fmt.format(bar=bar, barlen=barlen, nfill=nfill, size=size / factor)
+            print(val, end="", flush=True)
+
+    return _update
```

Comment:
Simplification opportunity: we could replace this with the following snippet (or entirely delete the function -- it's just telling us about our current progress)

```python
 def _progress_bar(tot_bytes, *, silent=False):
    """provides information about the current download"""
    power_div_3 = int(log10(tot_bytes) // 3) if tot_bytes > 0 else 0
    div, unit = 1000.0**power_div_3, (" B", "KB", "MB", "GB")[power_div_3]
    done = tot_bytes <= 0 or silent or not sys.stdout.isatty()

    def _update(size):
        nonlocal done
        m = "\n" if size is None else f"\r{size / div:.2f}/{tot_bytes / div:.2f} {unit}"
        if not done:
            print(m, end="", flush=True)
            done = size is None

    return _update